src/package-request.js

```diff
@@ -281,7 +281,8 @@ export default class PackageRequest {
       promises.push(this.resolver.find({
         pattern: depPattern,
         registry: remote.registry,
-        optional: false,
+        // dependencies of optional dependencies should themselves be optional
+        optional: this.optional,
```

Comment:
Hmm...couldn't your scenario be simplified to

```
(optional) A -> C
B -> C
```

and still prove what you are concerned about? I'm not seeing why the first chain has to have a depth of 3 -- am I missing something?